### ISSUE TITLE
Question about `feat.transpose(2, 3)` in `index_2d` before `grid_sample`

### ISSUE THREAD

**Issue by Ishak96** [xmed-lab/DIF-Net#18]
Hi 👋
I have a question regarding the implementation of the following function, specifically the transpose applied to `feat`:

```python
def index_2d(feat, uv):
    # feat: [B, C, H, W]
    # uv: [B, N, 2]
    uv = uv.unsqueeze(2)  # [B, N, 1, 2]
    feat = feat.transpose(2, 3)  # [B, C, W, H]
    samples = torch.nn.functional.grid_sample(
        feat, uv, align_corners=True
    )  # [B, C, N, 1]
    return samples[:, :, :, 0]  # [B, C, N]
```

According to the PyTorch documentation, `grid_sample` expects the input feature map to be in `[B, C, H, W]` format, where the grid’s last dimension corresponds to `(x, y)` indexing over `(W, H)`.

Transposing `feat` from `[B, C, H, W]` to `[B, C, W, H]` appears to swap the spatial axes. This makes it unclear whether:

* the transpose is strictly necessary,
* the `uv` coordinates are assumed to be in `(y, x)` order instead of `(x, y)`

Could you please clarify the rationale behind this transpose, or whether it can be safely removed if `uv` is defined in standard `(x, y)` normalized coordinates?

Thanks!

https://github.com/xmed-lab/DIF-Net/blob/ddc4c3dbe6990d8898d7114e423d86458b98b900/code/models/model.py#L16C29-L16C30

**Comment by lyqun** [comment]
The transpose is not necessary if the uv is ordered in the same way as last two dimensions of feat.

You can write a quick example to check if your data (uv, feat) is aligned in their orders. For example, create a image (as feat) with a shape of [1, 3, 256, 256] to see if the result is correct with aligned uv and no transpose.

You may refer to the doumentation of this function for more details: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html

**Comment by Ishak96** [comment]
Perfect, thank you! One more question regarding training.

I’m trying to reproduce your results on a different chest CT dataset (not LUNA16). I’m using the same configuration file provided in your C²RV paper, and I’m simulating the projections using ODL (Operator Discretization Library). Unfortunately, I’m not getting good results the training does not seem to converge.

Are there any additional adjustments or considerations I should take into account when switching datasets or projection operators?
Also, is it necessary to train on a single GPU to reproduce the reported results?

Thanks again for your help!

**Comment by lyqun** [comment]
1. If the performance is fair (not bad), you may consider using more data for training or improving the image quality of simulated projections (e.g., modify geometry parameters).
2. If the performance is terrible (nothing correct is reconstructed), you need to check if the indexing function (to query view-specific features of a 3d point from 2d projections) is correctly used.

**Comment by Ishak96** [comment]
So training on 8 GPUs using a DDP strategy is not a problem, correct?

For reference, here is an example of the simulated projections generated using ODL (this is a publicly available dataset), with 256 detectors.

<img width="696" height="692" alt="Image" src="https://github.com/user-attachments/assets/c1fcb7ff-c268-4be5-93b7-09c8f558f9c8" />

**Comment by Ishak96** [comment]
The performance I’m getting is between fair and poor. DIF-Net is able to recover the overall shape, but the reconstructed intensities are completely off very bright, similar to what you would expect when reconstructing with the adjoint of the Radon.

**Comment by lyqun** [comment]
Using 8 gpus is fine. You may try training with more iterations. If you are using this repo, using 8 gpus means fewer iterations and the model may not converge well. btw, please check the geometry first. I am not very sure if the geometry setting is consistent if you are using this repo for training while using c2rv for data processing.

The simulated projections seem to be good.

**Comment by Ishak96** [comment]
Hi Yiqun,
After moving from ODL to TIGRE, everything is working correctly. Thank you very much for your help.
